dimod/discrete/discrete_quadratic_model.py

```diff
@@ -810,3 +808,315 @@ def to_numpy_vectors(self, return_offset: bool = False):
+                raise ValueError(f'unknown case label {v}')
+
+            v_m = self._unique_label_case[v]
+            return DQM.get_quadratic_case(self, *u_k, *v_m)
+        else:
+            return DQM.get_quadratic(self, u, v)
+
+    def get_quadratic_case(self, u, u_case, v, v_case):
+        """The bias associated with the interaction between two cases of `u`
+        and `v`.
+
+        Args:
+            u: A variable in the discrete quadratic model.
+
+            u_case: The case of `u`.
+
+            v: A variable in the discrete quadratic model.
+
+            v_case: The case of `v`.
+
+        Returns:
+            The quadratic bias.
+
+        """
+        u_case = self._lookup_shared_case(u, u_case)
+        v_case = self._lookup_shared_case(v, v_case)
+        return DQM.get_quadratic_case(self, u, u_case, v, v_case)
+
+    def set_linear(self, v, biases):
+        """Set the linear bias(es) associated with `v`.
+
+        Args:
+            v: A variable in the discrete quadratic model, or a unique case
+                label.
+
+            biases (float or array-like):  If `v` is a variable, the linear
+                biases in an array.  Otherwise, the linear bias is a real
+                number.
+
+        """
+        v_k = self._unique_label_case.get(v)
+        if v_k:
+            DQM.set_linear_case(self, *v_k, biases)
+        else:
+            DQM.set_linear(self, v, biases)
+
+    def set_linear_case(self, v, case, bias):
+        """The linear bias associated with case `case` of variable `v`.
+
+        Args:
+            v: A variable in the discrete quadratic model.
+
+            case: The case of `v`.
+
+            bias (float): The linear bias.
+
+        """
+        case = self._lookup_shared_case(v, case)
+        DQM.set_linear_case(self, v, case, bias)
+
+    def set_quadratic(self, u, v, biases):
+        """Set bias(es) associated with the interaction between `u` and `v`.
+
+        Args:
+            u: A variable in the discrete quadratic model, or a unique case
+                label.  If `u` is a unique case label, `v` must be a unique
+                case label.
+
+            v: A variable in the discrete quadratic model, or a unique case
+                label.  If `u` is a unique case label, `v` must be a unique
+                case label.
+
+            biases (float or array-like/dict):
+                The quadratic bias(es).  If `u` and `v` are variables, then
+                `biases` may be a dictionary of the form
+                `{case_u, case_v: bias, ...}` or a
+                :meth:`~DiscreteQuadraticModel.num_cases(u)` by
+                :meth:`~DiscreteQuadraticModel.num_cases(v)` array-like.
+
+                If `u` and `v` are unique case labels, the quadratic bias is a
+                real number.
+
+        Raises:
+            ValueError: If `u` is a unique case label and `v` is not.
+
+        """
+        u_k = self._unique_label_case.get(u)
+        if u_k:
```

Comment:
unless I am missing something, this can be Falsey for lots of cases. E.g. if the label is `''` or `0`.

Reply:
🤦 You're right, of course!

Reply:
Leaving aside the code and considering the design space, is there a good reason to support `''` or non-strings as labels?  Supporting non-empty strings only, as labels, covers both example use cases given in #840.  The "labels can be any hashable object" design can be revised.

Reply:
Frozensets and tuples are both useful in some cases. And if we're going to use strings, why not include `''`? And in general, why restrict the user? Hashable is well defined in python. For what it's worth, there is actually one restriction, which is that we don't allow `None`, which is hashable. In this case I think you could just do `if u_k is not None`.

Reply:
Update: `u_k` will be either a 2-tuple - a (var, case) pair - or `None`.  So this code is correct.  I do need to add a check for case labels being `None` to `add_variable()` though.

Reply:
Hmm, or perhaps better yet, we could allow `None` as a case label.

Reply:
We could for sure. FWIW, the reason that we don't allow `None` as a variable label is so that we can allow labels as optional keyword arguments easily. E.g. `variables._append(label=None)` does not mean that the variable should be labelled `None`, but rather that a new label should be generated. It might make sense to reserve that case for the future, or not :shrug: